flyteadmin/pkg/common/filters.go

```diff
@@ -389,3 +391,62 @@ func NewWithDefaultValueFilter(defaultValue interface{}, filter InlineFilter) (I
 		defaultValue:     defaultValue,
 	}, nil
 }
+
+// ResourceColumns is a struct to indicate which columns in a query represent the flyte project and domain
+type ResourceColumns struct {
+	Project string
+	Domain  string
+}
+
+// IsolationFilter is an interface for filtering data based on authorization rules
+type IsolationFilter interface {
+	GetScopes() []func(tx *gorm.DB) *gorm.DB
+}
+
+// ResourceIsolationFilter is an implementation of IsolationFilter that provides db scopes for filtering resources.
+type ResourceIsolationFilter struct {
+	resourceScopes []isolation.ResourceScope
+	columns        ResourceColumns
+}
+
+// NewResourceIsolationFilter creates a new ResourceIsolationFilter
+func NewResourceIsolationFilter(resourceScopes []isolation.ResourceScope, columns ResourceColumns) *ResourceIsolationFilter {
+	return &ResourceIsolationFilter{resourceScopes: resourceScopes, columns: columns}
+}
+
+// GetScopes gets a list of functions that mutates a query to apply filtering where clauses
+func (a *ResourceIsolationFilter) GetScopes() []func(tx *gorm.DB) *gorm.DB {
+	scopes := []func(tx *gorm.DB) *gorm.DB{}
+
+	for tempIndex, tempResourceScope := range a.resourceScopes {
+		// Copy to avoid issues
+		index := tempIndex
+		resourceScope := tempResourceScope
+		scopes = append(scopes, func(tx *gorm.DB) *gorm.DB {
+			// We must start of with a where clause, then after that we can chain each resource scope with OR
+			if index == 0 {
+				return tx.Where(tx.Scopes(a.toDbScopes(resourceScope)))
+			}
+			return tx.Or(tx.Scopes(a.toDbScopes(resourceScope)))
+		})
+	}
+
+	return scopes
+}
+
+func (a *ResourceIsolationFilter) toDbScopes(resourceScope isolation.ResourceScope) func(db *gorm.DB) *gorm.DB {
```

Comment:
One key implementation detail is that we use GORM's scopes in order to dynamically alter queries: https://gorm.io/docs/advanced_query.html#Scopes